src/inpod/statemanager.rs

```diff
@@ -0,0 +1,522 @@
+        let handle = tokio::spawn(workload_state.drain.drain());
+        // before we push to draining, try to clear done entries, so the vector doesn't grow too much
+        self.draining.retain(|x| !x.is_finished());
+        // add deleted pod to draining. we do this so we make sure to wait for it incase we
+        // get the global drain signal.
+        self.draining.push(handle);
+    }
+
+    async fn join(self) {
+        futures::future::join_all(self.draining).await;
+    }
+}
+
+pub struct WorkloadProxyManagerState {
+    proxy_gen: ProxyFactory,
+    metrics: Arc<Metrics>,
+    // use hashbrown for extract_if
+    workload_states: hashbrown::HashMap<String, WorkloadState>,
+
+    // workloads we wanted to start but couldn't because we had an error starting them.
+    // This happened to use mainly in testing when we redeploy ztunnel, and the old pod was
+    // not completely drained yet.
+    pending_workloads: hashbrown::HashMap<String, (WorkloadInfo, InpodNetns)>,
+    draining: DrainingTasks,
+
+    // new connection stuff
+    snapshot_received: bool,
+    snapshot_names: std::collections::HashSet<String>,
+
+    inpod_config: InPodConfig,
+}
+
+impl WorkloadProxyManagerState {
+    pub fn new(proxy_gen: ProxyFactory, inpod_config: InPodConfig, metrics: Arc<Metrics>) -> Self {
+        WorkloadProxyManagerState {
+            proxy_gen,
+            metrics,
+            workload_states: Default::default(),
+            pending_workloads: Default::default(),
+            draining: Default::default(),
+
+            snapshot_received: false,
+            snapshot_names: Default::default(),
+            inpod_config,
+        }
+    }
+
+    #[cfg(test)] // only used in tests, so added this to avoid warning
+    pub(super) fn workload_states(&self) -> &hashbrown::HashMap<String, WorkloadState> {
+        &self.workload_states
+    }
+
+    // Call this on new connection
+    pub fn reset_snapshot(&mut self) {
+        self.snapshot_names.clear();
+        self.pending_workloads.clear();
+        self.snapshot_received = false;
+    }
+
+    pub async fn process_msg(&mut self, msg: WorkloadMessage) -> Result<(), Error> {
+        match msg {
+            WorkloadMessage::AddWorkload(poddata) => {
+                info!(
+                    "pod {} received netns, starting proxy",
+                    poddata.info.workload_uid
+                );
+                if !self.snapshot_received {
+                    self.snapshot_names
+                        .insert(poddata.info.workload_uid.clone());
+                }
+                let netns = InpodNetns::new(self.inpod_config.cur_netns(), poddata.netns)
+                    .map_err(|e| Error::ProxyError(crate::proxy::Error::Io(e)))?;
+
+                self.add_workload(poddata.info, netns)
+                    .await
+                    .map_err(Error::ProxyError)
+            }
+            WorkloadMessage::KeepWorkload(workload_uid) => {
```

Comment:
I don't feel very comfortable with this KeepWorkload thing.

Either our netns logic is reliable and we don't need this, or its unreliable and we need to abandon this project or make it more reliable.

(I think we can make it reliable)

Reply:
I agree with the sentiment. This is my thought process that led me to add this:

The current approach we have to restore state when the node agent restarts is "eventually reliable",  so this is here to avoid temporary blips when things restart in edge cases. Let me explain in more detail:

The current method to get the FD after node agent restart is by scanning the /proc file system.

I believe this leaves us with a theoretical race:
If the node agent restarts after the CNI was invoked, but before processes belonging to the pod were created. In this case, the workload is in the ztunnel but not yet available in the node agent.

When ztunnel (re)connects to node agent, it gets a "snapshot" of workloads, so it can remove workloads that may have been removed while it was disconnected. So if that happens at the same time, the ztunnel might think that a workload was removed when it wasn't..

While extremely unlikely, i figured better safe than sorry.

Reply:
We shouldn't be using processes in the users containers, we can utilize the pause container? That should run before CNI is invoked and never exit I believe?

Reply:
re @howardjohn good question about the pause container - we'll need to look if containerd and cri-o start it before they call the CNI. Assuming it checks out, If we remove this from here, are you comfortable placing a `panic()`  in the node agent (where it would otherwise send the keep-workload) so it is loud and clear if we are wrong?

Reply:
re @costinm [comment here](https://github.com/istio/ztunnel/pull/747#issuecomment-1876096537) 

> The problem is that 'complex' is rarely safe - and most of the time sorry.

I'm aware.

> and not at all clear what 'remove' is supposed to do, the pod will be deleted anyways

Let me clarify.

When a pod is deleted, its network namespace doesn't get deleted. That's because network namespaces in linux can't be explicitly deleted. They automatically get cleaned up when nothing references them. Therefore, ztunnel needs some mechanism to know that a pod is deleted, so it can clean up its resources.

Our current approach, is that the node-agent tells ztunnel that pods were deleted. And there for, if the ztunnel was disconnected from the the node agent, when it reconnects we need to reconcile the full state of local pods on the node, to account for pods that may have been deleted while the ztunnel was disconnected.

We do this by sending a bunch of AddWorkload messages, followed by a SnapshotSent message on the initial connection. The ztunnel will clean up and workloads that it didn't see in one of these WorkloadAdded messages sent prior to the SnapshotSent message.
Aftet the SnapshotSent message, the protocol becomes delta-like with AddWorkload \ DelWorkload sent as needed.

If/When the node agent restarts, it needs to reconstruct its local state. That state is composed of local pods -> netns mapping. The node agent uses heuristics to reconstruct said state.

The (potentially non-existent [see above]) problem that KeepWorkload aims to solve, is to account for a possible temporary issue with that heuristic. Thus allowing the node agent to tell the ztunnel when it reconnects, that even though it doesn't have the netns for a certain pod, the ztunnel shouldn't remove it after the SnapshotSent message is sent.